lib/Sema/ConstraintSystem.cpp

```diff
@@ -513,6 +513,13 @@ ConstraintLocator *ConstraintSystem::getCalleeLocator(
                                   {LocatorPathElt::ApplyFunction(),
                                    LocatorPathElt::ImplicitCallAsFunction()});
     }
+
+    // Handling an apply for a nominal type that supports @dynamicCallable.
+    auto nominal = fnTy->getAnyNominal();
+    if (nominal && nominal->getAttrs().hasAttribute<DynamicCallableAttr>()) {
```

Comment:
It would be great if we could also handle the other cases where `@dynamicCallable` is applicable, e.g through protocol compositions and superclasses. You should be able to check this through calling `getDynamicCallableMethods` and checking if the result `isValid()` (though it would need to become a `ConstraintSystem` method first).

If you want, it would also be great to turn `getDynamicCallableMethods` into a request such that the results can be cached across multiple constraint systems, similar to `HasDynamicMemberLookupAttributeRequest`.

Reply:
Thank you for the review @hamishknight :) Already updated 👍